GG/src/RichText/ImageBlock.cpp

```diff
@@ -0,0 +1,150 @@
+#include <GG/RichText/ImageBlock.h>
+
+#include <GG/Texture.h>
+#include <GG/RichText/RichText.h>
+#include <GG/TextControl.h>
+#include <GG/DrawUtil.h>
+
+#include <boost/filesystem.hpp>
+#include <algorithm>
+
+namespace GG {
+
+    const std::string ImageBlock::IMAGE_TAG ("img");
+
+    ImageBlock::ImageBlock (const std::string& image_path, X x, Y y, X w, GG::Flags< GG::WndFlag > flags) :
+        BlockControl (x, y, w, flags),
+        m_graphic (0) {
+        try {
+            boost::shared_ptr<Texture> texture = GetTextureManager().GetTexture (image_path);
+            m_graphic = new StaticGraphic (texture, GRAPHIC_PROPSCALE | GRAPHIC_SHRINKFIT | GRAPHIC_CENTER);
+            AttachChild (m_graphic);
+        } catch (GG::Texture::BadFile& errBadFile) {
+            // No can do inside GiGi.
+        }
+    }
+
+    Pt ImageBlock::SetMaxWidth (X width) {
+
+        if (m_graphic != 0) {
+            // Give the graphic the set width and give it liberty with the height.
+            m_graphic->Resize (Pt (width, Y (INT_MAX)));
+
+            // Get the actual space the graphic decided to use.
+            Rect rect = m_graphic->RenderedArea();
+            Pt size = rect.LowerRight() - rect.UpperLeft();
+
+            // Take up the full width to center the image.
+            size.x = width;
+
+            // Don't take the extra vertical space.
+            m_graphic->Resize (size);
+
+            // Update our size to match the graphic we are wrapping.
+            Resize (size);
+
+            // Return the size we decided to be.
+            return size;
+        } else {
+            // We don't have an image. Take a quarter of width to show an X.
+            Pt size (width, Y (Value (width) / 4));
+            Resize (size);
+            return size;
+        }
+    }
+
+    void ImageBlock::Render() {
+        if (m_graphic == 0) {
+            // Error: no image. Draw red x.
+            Pt ul = UpperLeft();
+            Pt lr = LowerRight();
+            Pt size = lr - ul;
+            ul.x += size.x / 2 - X (Value (size.y)) / 2;
+            lr.x -= size.x / 2 - X (Value (size.y)) / 2;
+            FlatX (ul, lr, CLR_RED);
```

Comment:
FlatX is at GG/src/DrawUtil.cpp:751.

The missing.png is a FreeOrion asset, ImageBlock is in GG. Besides, relying on an external resource to indicate that you are failing to find an external resource is slightly fishy in my opinion.